invalid ebpf in example_sock_test.go

**Describe the bug**
If I execute  `Example_socket()`   from example_sock_test.go the validator reject the ebpf program. I compared the ebpf with that of "[Linux/samples/bpf/sock_example.c](https://github.com/torvalds/linux/blob/master/samples/bpf/sock_example.c)" and noticed some difference. After modifying the ebpf to match that of sample program from Linux kernel I got the program to work. I suspect this is because of incorrect generated ebpf, don't think it's because of different setup (kernel, ..). But if it work's on you computer it might be.

**To Reproduce**
If I execute `Example_socket()` i get: 
```
         0: MovReg dst: r6 src: r1
         1: LdAbsB imm: 25
         2: StXMemW dst: rfp src: r0 off: -4 imm: 0
         3: MovReg dst: r2 src: rfp
         4: SubImm dst: r2 imm: 4
         5: LdImmDW dst: r1 imm: 3
         7: Call MapLookupElement
         8: JEqImm dst: r0 off: -1 imm: 0 <out>
         9: MovImm dst: r1 imm: 1
        10: INVALID
out:
        11: MovImm dst: r0 imm: 0
        12: Exit

panic: invalid operation at position 9
```
changed xadd 10 instruction to the one from kernel
```
opcode := asm.OpCode(asm.StXClass) | asm.OpCode(asm.XAddMode).SetSize(asm.DWord)
newXAdd := asm.XAdd(asm.R0, asm.R1)
newXAdd.OpCode = opcode 
```
Now i got a error about subtraction from stack pointer 
changed substraction in addition of negative number  as in kernel

```
         0: MovReg dst: r6 src: r1
         1: LdAbsB imm: 25
         2: StXMemW dst: rfp src: r0 off: -4 imm: 0
         3: MovReg dst: r2 src: rfp
         4: SubImm dst: r2 imm: 4
         5: LdImmDW dst: r1 imm: 3
         7: Call MapLookupElement
         8: JEqImm dst: r0 off: -1 imm: 0 <out>
         9: MovImm dst: r1 imm: 1
        10: StXXAddDW dst: r0 src: r1
out:
        11: MovImm dst: r0 imm: 0
        12: Exit

panic: failed to load program: permission denied: 0: (bf) r6 = r1
1: (30) r0 = *(u8 *)skb[25]
2: (63) *(u32 *)(r10 -4) = r0
3: (bf) r2 = r10
4: (17) r2 -= 4
R2 subtraction from stack pointer prohibited
```
Now i got error  that ` 5: LdImmDW dst: r1 imm: 3` was not a map_ptr
(3 was file descriptor)
```   
         0: MovReg dst: r6 src: r1
         1: LdAbsB imm: 25
         2: StXMemW dst: rfp src: r0 off: -4 imm: 0
         3: MovReg dst: r2 src: rfp
         4: AddImm dst: r2 imm: -4
         5: LdImmDW dst: r1 imm: 3
         7: Call MapLookupElement
         8: JEqImm dst: r0 off: -1 imm: 0 <out>
         9: MovImm dst: r1 imm: 1
        10: StXXAddDW dst: r0 src: r1
out:
        11: MovImm dst: r0 imm: 0
        12: Exit

panic: failed to load program: permission denied: 0: (bf) r6 = r1
1: (30) r0 = *(u8 *)skb[25]
2: (63) *(u32 *)(r10 -4) = r0
3: (bf) r2 = r10
4: (07) r2 += -4
5: (18) r1 = 0x3
7: (85) call bpf_map_lookup_elem#1
R1 type=inv expected=map_ptr

```
Linux kernel use separate marco BPF_LD_MAP to load map, 
found not much in there documentation about this, but in the source code its defined as:
```
#ifndef BPF_PSEUDO_MAP_FD
# define BPF_PSEUDO_MAP_FD  1
#endif

/* pseudo BPF_LD_IMM64 insn used to refer to process-local map_fd */
#define BPF_LD_MAP_FD(DST, MAP_FD)              \
    BPF_LD_IMM64_RAW(DST, BPF_PSEUDO_MAP_FD, MAP_FD)
```
so changed the load with:
```
loadMap := asm.LoadImm(asm.R1, int64(mapFd), asm.DWord)
loadMap.Src = 1 
```
Now the programs runs successful, but output is incorrect.
This was because  `unsafe.Offsetof(ip.Protocol))` was incorrect because 
IPHrd has no flags filed 
```
type IPHdr struct {
      VersionIHL uint8
      Tos        uint8                                                                                                                                                                                                                                                           
      Length     uint16
      ID         uint16
      // Flags          uint8
      FragmentOffset uint16
      TTL            uint8
      Protocol       uint8
      Checksum       uint16
      SrcIP          [4]byte
      DestIP         [4]byte
      PayloadLength  int
}
```
Now it worked for me.

**Linux Information **
 - Version : `Linux Arch 4.19.12-arch1-1-ARCH #1 SMP PREEMPT Fri Dec 21 13:56:54 UTC 2018 x86_64 GNU/Linux` ( had same problems with 4.14 but have not test that extensive)

 - Distribution: Arch linux 

**Additional **
I like the project, and hope this issue Helps. Hope it was clear.  
If you want want, I can send pull request for:
- change opcode for xadd
- add LoadMap instruction ( probaly should also be changed in example_sock_extract_dist_test.go)
- change example_sock_test.go 
     * change   ` asm.Sub.Imm(asm.R2, 4),`  ->   ` asm.Add.Imm(asm.R2, -4),` 
     * remove IPHdr.Flags

--- a/asm/load_store.go
+++ b/asm/load_store.go
@@ -157,14 +157,14 @@ func StoreImm(dst Register, offset int16, value int64, size Size) Instruction {
 }
 
 // XAddOp returns the OpCode to atomically add a register to a value in memory.
-func XAddOp() OpCode {
-	return OpCode(StClass).SetMode(XAddMode).SetSource(RegSource)
+func XAddOp(size Size) OpCode {
+	return OpCode(StXClass).SetMode(XAddMode).SetSize(size)
 }
 
 // XAdd atomically adds src to *dst.
-func XAdd(dst, src Register) Instruction {
+func XAdd(dst, src Register, size Size) Instruction {
 	return Instruction{
-		OpCode: XAddOp(),
+		OpCode: XAddOp(size),
 		Dst:    dst,
 		Src:    src,
 	}

--- a/example_sock_extract_dist_test.go
+++ b/example_sock_extract_dist_test.go
@@ -128,7 +128,7 @@ func attachBPF(fd int) (*ebpf.Map, error) {
 		// load ok? inc. Otherwise? jmp to mapupdate
 		asm.JEq.Imm(asm.R0, 0, "update-map"),
 		asm.Mov.Imm(asm.R1, 1),
-		asm.XAdd(asm.R0, asm.R1),
+		asm.XAdd(asm.R0, asm.R1, asm.DWord),
 		asm.Ja.Label("exit"),
 
 		// MapUpdate

--- a/example_sock_test.go
+++ b/example_sock_test.go
@@ -76,7 +76,7 @@ func Example_socket() {
 		asm.Mov.Imm(asm.R1, 1),
 		// atomically increment register
 		// xaddst r0, r1
-		asm.XAdd(asm.R0, asm.R1),
+		asm.XAdd(asm.R0, asm.R1, asm.DWord),
 		// set exit code to 0
 		// mov r0, imm
 		asm.Mov.Imm(asm.R0, 0).Sym("out"),

